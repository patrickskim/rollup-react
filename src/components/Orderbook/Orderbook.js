import React from 'react';
import PropTypes from 'react-proptypes';
import classNames from 'classnames/bind';
import styles from './Orderbook.module.scss'

const cx = classNames.bind(styles);

console.log('another', styles)

// TODO better name
const calc = ({ price, quantity }) => ({ });

const Row = (props)  => {
  const { price, quantity, total } = props;
  return (
    <tr key={price}>
      <td>{fmtUSD(price)}</td>
      <td>{fmtBTC(quantity)}</td>
      <td>{fmt$(total)}</td>
    </tr>
  );
};

// TODO generalize fmt
const fmt$ = (num) => new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(num);
const fmtBTC = (num) => new Intl.NumberFormat('en-US', { style: 'decimal', minimumFractionDigits: 6, minimumIntegerDigits: 1 }).format(num);
const fmtUSD = (num) => new Intl.NumberFormat('en-US', { style: 'decimal', minimumFractionDigits: 2, maximumFractionDigits: 2 }).format(num);

const Orderbook = (props) => {
  const {
    orders,
    viewLimit,
  } = props;

  function renderRows(orders, viewLimit = 5) {
    let rows = [];
    let prev = {};

    orders.forEach((order, idx) => {
      const { price, quantity } = order;
      const total = price * quantity

      if (idx == 0) {
        prev = { price, quantity, total };

        return false;
      }

      if (prev.price != price) {
        rows.push(<Row {...prev} />);
        prev = { price, quantity, total };

        return false;
      }

      prev.quantity = prev.quantity + quantity;
      prev.total = prev.total + total;
    });

    return rows.slice(0, viewLimit);
  }

  function calcAvg(orders) {
    const { bids, asks } = orders;
    const lowest = asks[0];
    const highest = bids[0];

    return (lowest.price + highest.price) / 2;
  }

  const avg = calcAvg(orders);
  const bids = renderRows(orders.bids);
  const asks = renderRows(orders.asks);

  return (
    <div className={styles.Sample}>
      <thead>
        <th>Price (USDT)</th>
        <th>Amount (BTC)</th>
        <th>Total</th>
      </thead>
      <tbody className={styles.bids}>{bids}</tbody>
      <tr className={styles.avg}>
        <td colspan={3}>{fmtUSD(avg)}</td>
      </tr>
      <tbody className={styles.asks}>{asks}</tbody>
    </div>
  );
};

Orderbook.propTypes = {
  orders: PropTypes.shape({
    asks: PropTypes.arrayOf(
      PropTypes.shape({
        quantity: PropTypes.number.isRequired,
        price: PropTypes.number.isRequired
      })
    ).isRequired,
    bids: PropTypes.arrayOf(
      PropTypes.shape({
        quantity: PropTypes.number.isRequired,
        price: PropTypes.number.isRequired
      })
    ).isRequired,
  }),
  viewLimit: PropTypes.number
};

export default Orderbook;
