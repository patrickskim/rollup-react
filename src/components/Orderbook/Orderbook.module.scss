.Sample {
  font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
  font-size: 12px;
  line-height: 1em;

  tbody {
    position: relative;
    margin: 10px 0;
  }

  th {
    background: red;
    padding: 15px 0;
  }

  td {
    position: relative;
    padding: 7px 0;
    text-align: center;

    &:first-child {
      text-align: left;
    }

    &:last-child {
      color: blue;
      text-align: right;
    }
  }

  .asks td:first-child {
    color: green;
  }

  .bids td:first-child  {
    color: red;
  }

  .avg {
    font-size: 20px;
    text-align: center;

    td {
      padding: 15px 0;

      font-size: 20px;
      text-align: center;

      border: 2px solid #eee;
      border-left: 0;
      border-right: 0;
    }
  }
}



