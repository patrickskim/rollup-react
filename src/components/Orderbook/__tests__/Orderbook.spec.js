import React from 'react';
import Orderbook from '../.';
import TestRenderer from 'react-test-renderer';

describe('<Orderbook />', () => {
  const props = {
    orders: {
      bids: [{
        price: 0,
        quantity: 1,
      }],
      asks: [{
        price: 0,
        quantity: 1,
      }]
    }
  };

  it('renders correctly', () => {
    const actual = TestRenderer.create(<Orderbook {...props} />).toJSON();

    expect(actual).toMatchSnapshot();
  });

});
