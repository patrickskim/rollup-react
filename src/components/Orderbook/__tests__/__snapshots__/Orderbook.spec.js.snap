// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Orderbook /> renders correctly 1`] = `
<div
  className="Sample"
>
  <thead>
    <th>
      Price (USDT)
    </th>
    <th>
      Amount (BTC)
    </th>
    <th>
      Total
    </th>
  </thead>
  <tbody
    className="bids"
  />
  <tr
    className="avg"
  >
    <td
      colspan={3}
    >
      0.00
    </td>
  </tr>
  <tbody
    className="asks"
  />
</div>
`;
